src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,40 @@
+package mate.academy.service.impl;
+
+import java.util.Optional;
+import mate.academy.exception.AuthenticationException;
+import mate.academy.lib.Inject;
+import mate.academy.lib.Service;
+import mate.academy.model.User;
+import mate.academy.service.AuthenticationService;
+import mate.academy.service.UserService;
+import mate.academy.util.HashUtilSha512;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> userByEmail = userService.findByEmail(email);
+        if (userByEmail.isPresent()) {
+            User user = userByEmail.get();
+            String hashedPassword = HashUtilSha512.hashPassword(password, user.getSalt());
+            if (user.getPassword().equals(hashedPassword)) {
+                return user;
+            }
+        }
+        throw new AuthenticationException("Can't authenticate user email: " + email);
+    }
+
+    @Override
+    public User register(String email, String password) {
+        if (userService.findByEmail(email).isPresent()) {
+            throw new RuntimeException("This email " + email + " is already registered");
```

Comment:
It is better to throw custom RegistrationException and it's also shouldn't extend RuntimeException

Reply:
👍